README.md
# Spritz.js

📽Modern and delightful **sprites animation library** for JavaScript!

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/maoosi/spritz.js/master/LICENSE.md) [![GitHub release](https://img.shields.io/github/release/maoosi/spritz.js.svg)](https://github.com/maoosi/spritz.js) [![GitHub issues](https://img.shields.io/github/issues/maoosi/spritz.js.svg)](https://github.com/maoosi/spritz.js/issues) [![Travis CI](https://travis-ci.org/maoosi/spritz.js.svg?branch=master)](https://travis-ci.org/maoosi/spritz.js)

*Can be used for interactive animations, 360 viewers, HTML5 games, and more...*

> ✨ **Try the [**Demo samples**](http://codepen.io/collection/XQZjMx/).**

![](spritz.cover.gif)

## Features

* **Lightweight** - under 3.3KB minified and gzipped.
* **Responsive** - syntax following the new HTML5 `picture` element.
* **Modern** - written in ES6 JavaScript, no jQuery required.
* **[Compatible](#browser-support)** - IE10+ support, mobile support.
* **[Chainable API](#api)** - play, pause, wait, and more...

## Getting Started

1. [Installation](#installation)
2. [Usage](#usage)
3. [Options](#options)
4. [API / Methods](#api)
5. [Events](#events)
6. [Browser Support](#browser-support)
7. [Contribute](#contribute)
8. [Projects using it!](#projects-using-it)
9. [License](#license)

## Installation

### Using Yarn

```bash
yarn add spritz.js
```

### Using NPM

```bash
npm i spritz.js --save
```

### Unpkg CDN

```html
<script src="https://unpkg.com/spritz.js@2.3.0/dist/spritz.js"></script>
```

## Usage

The sprite will be generated inside the HTML element passed as first parameter. The canvas dimensions will depend on the width and the height of this element.

```html
<div id="#selector"></div>
```

```javascript
// import the library
import Spritz from 'spritz.js'

// instantiate with options
const sprite = Spritz('#selector', { /* options here */ })

// 🎉 ready to control your sprite!
sprite.play().wait(2000).pause()
```

## Options

Settings overview list:

```javascript
Spritz('#selector', {
	picture: [],
	steps: 1,
	rows: 1
})
```

Full option details are detailed below.

| Option | Type | Default | Description |
| --- | --- | --- | --- |
| `picture` | *array* | `[]` | Array of picture objects to be used as Sprite. The syntax is close to the new HTML5 `picture` element. [More details below.](#picture) |
| `steps` | *integer* | `1` | Number of steps (or frames) composing the sprite. |
| `rows` | *integer* | `1` | Number of rows (or lines) composing the sprite. |
| `init` | *integer* | `1` | Initial sprite step. Set to `false` to disable the default auto‑init() feature. |

Sprite object will be displayed as per background-size .

### Picture

The picture option syntax is close to the new HTML5 `picture` element. The library will choose the most suitable source according to the current layout of the page and the browser it will be displayed into. Sense of reading is following the array order (first to last element).

**Parameters:**

| Key | Type | Description |
| --- | --- | --- |
| `srcset` | required | Accepts a single image file path (e.g. `srcset: 'kitten.png'`). WebP format accepted for compatible browsers (built-in detection). |
| `media ` | optional | Accepts any valid media query that you would normally find in a CSS @media selector (e.g. `media: '(max-width: 1200px)'`). |
| `width ` | required | Specify the full width in pixels, of the sprite source image (e.g. `width: 7800`). |
| `height ` | required | Specify the full width in pixels, of the sprite source image (e.g. `height: 2829`). |
| `objectFit` | optional | Specifies how the sprite should be fitted to the parent. Values: `contain` or `cover`. Default: `contain`. |

**Example:**

```javascript
[
	{
		srcset: 'kitten-highres.png',
		media: '(min-width: 1200px)',
		objectFit: 'cover',
		width: 7800, height: 2829
	},
	{
		srcset: 'kitten.webp',
		width: 3900, height: 1415
	},
	{
		srcset: 'kitten.png',
		width: 3900, height: 1415
	}
]
```

## API

Spritz exposes the following chainable API methods.

### Basic methods

* [fps](#fps)
* [play](#play)
* [playback](#playback)
* [pause](#pause)
* [stop](#stop)
* [wait](#wait)
* [step](#step)
* [next](#next)
* [prev](#prev)

### Advanced methods

* [init](#init)
* [destroy](#destroy)
* [until](#until)
* [flip](#flip)
* [get](#get)

#### .fps()

Change animation speed from its default value (15).

```javascript
sprite.fps(10) // Change speed to 10fps
```

#### .play()

Play animation forward (using the current fps value). A `backward` option can be passed as parameter.

```javascript
sprite.play()
sprite.play('backward') // Alternative way to play an animation backward
```

#### .playback()

Play animation backward (using the current fps value).

```javascript
sprite.playback()
```

#### .pause()

Pause the current animation.

```javascript
sprite.pause()
```

#### .stop()

Stop the current animation (pause and reset to the initial step).

```javascript
sprite.stop()
```

#### .wait()

Chainable timeout that can be used to delay stuff. The delay value is to be passed as parameter, in milliseconds.

```javascript
sprite.wait(1000) // Wait for 1 second

/* Usage example */
sprite
	.play() // Play animation
	.wait(2000) // Then wait for 2 second
	.stop() // Then stop animation
```

#### .step()

Change the current step (or frame) of the sprite. Target step to be passed as parameter.

```javascript
sprite.step(5) // Change current step/frame to 5
```

#### .next()

Go to the next step (or frame).

```javascript
sprite.next()
```

#### .prev()

Go to the previous step (or frame).

```javascript
sprite.prev()
```

#### .init()

Build and load the sprite, within its selector. Initial step can be passed as parameter.

```javascript
sprite.init() // Basic usage
sprite.init(2) // Initial step is 2 (default = 1)
```

#### .destroy()

Completely destroy the sprite element and behaviors. Restore the initial state.

```javascript
sprite.destroy()
```

#### .until()

The next animation will automatically pause at the value specified. Two parameters can be used:

- step (required): Step/frame at which the animation should stop.
- loop (optional, default 1): Loop at which the animation should stop.

```javascript
sprite.until(7) // Animation will automatically stop at step/frame 7
sprite.until(3, 2) // Animation will automatically stop at step/frame 3, on the second loop

/* Usage example */
sprite
	.until(7) // Stop following animation at step/frame 7
	.play() // Play animation (will automatically stop at step/frame 7)
```

#### .flip()

Flip the sprite horizontally.

```javascript
sprite.flip()
```

#### .get()

Return data, then call the callback function with result. Two parameters can be used:

- data (required): Data to return. Possible values: `step`| `picture`.
- callback (optional): Callback function to be called with result as first parameter.

```javascript
// console log the current step/frame
sprite.get('step', (currentStep) => {
	console.log(currentStep)
})

// console log the current picture in use
sprite.get('picture', (pic) => {
	console.log(pic)
})
```

> Note that this method, like the others, is chainable.

## Events

The following events can be used to write custom callbacks functions.

```javascript
// Sprite is ready for use (after init)
sprite.on('ready', () => { /* Your code here */ })

// Sprite image has been loaded (current 'picture' loaded passed as parameter)
sprite.on('load', (picture) => { /* Your code here */ })

// Sprite has been destroyed
sprite.on('destroy', () => { /* Your code here */ })

// Sprite animation start ('forward' or 'backward' passed as first parameter)
sprite.on('play', (direction) => { /* Your code here */ })

// Sprite animation pause
sprite.on('pause', () => { /* Your code here */ })

// Sprite animation stop
sprite.on('stop', () => { /* Your code here */ })

// Animation timeout delay in  progress ('delay' value passed as parameter)
sprite.on('wait', (delay) => { /* Your code here */ })

// Step changed manually ('from' and 'to' passed as parameters)
sprite.on('change', (from, to) => { /* Your code here */ })

// Viewport has been resized (new current 'picture' passed as parameter)
sprite.on('resize', (picture) => { /* Your code here */ })
```

## Browser Support

Fully supported by Evergreen Browsers (Edge, Opera, Safari, Firefox & Chrome) and IE 10+ browsers.

> For **older browsers support like IE8 or IE9**, you'll need to manually include the following Polyfill library on your website:
>
```html
<script src="https://cdn.polyfill.io/v2/polyfill.min.js"></script>
```

## Contribute

```bash
yarn run watch
yarn run test
```

## License

[MIT](https://github.com/maoosi/spritz.js/blob/master/LICENSE.md) © 2017-2019 Sylvain Simao

[![forthebadge](http://forthebadge.com/images/badges/powered-by-electricity.svg)](http://forthebadge.com)


CHANGELOG.md
# Changelog


## v2.2.2

- Dependencies update


## v2.2.1

- Package published to Yarn
- Travis CI


## v2.2.0

- Change: `new Spritz().init()` combined into `Spritz()`.
- Bug fix: `.until()` was stoping the animation a step too early.
- New option `init`: Specifies if the sprite should be automatically initiated.
- Dev workflow moved from `Gulp` to `Rollup`.
- Added `CHANGELOG.md`.
- Dependency update: `wait.js@3.0.0`
- Dependency update: `knot.js@1.1.5`
- Documentation enhancement
- Unit testing with Jest


## v2.1.0

- Bug fix: sprite wrapper is now `position`:relative;`
- Bug fix: this access after the sprite has been destroyed
- Option objectFit moved to picture level


## v2.0.5

- Position bug fix


## v2.0.3

- Bug fix: blurry sprites
- Bug fix: calculation problems for certain sprites size
- New option objectFit: Specifies how the sprite should be fitted to the parent. Values: `contain` or `cover`


## v2.0.2

- Bug fix: stop() not stoping animations correctly
- Bug fix: blurry sprites


## v2.0.1

- Alternative way to play an animation backward using `.play('backward')`.
- Some `console.log()` removed.
- Solved bug: missing parameters for Events.


## v2.0.0

For this new version, the library has been completely rewritten from scratch. This means that version 2.x is not compatible with the version 1.x.

**Why rewriting everything from scratch?**

In order to propose better performances and support (by using a Canvas drawing approach instead of CSS only), images responsivity (by using a syntax close to the new HTML5 "picture" element), and more flexibility for designing animations (through a chainable API with timers).


## v1.0.0

- First release


test/test.js
import { strictEqual } from 'assert'
import Spritz from '../'
import jsdom from 'jsdom-global'


before(() => { jsdom() })
after(() => { jsdom() })


describe('Spritz.js', () => {


    // Create DOM
    let dom = [ { node: document.createElement('div'), class: 'sprite' } ]
    dom.forEach((el) => {
        document.body.appendChild(el.node)
        if (el.id) el.node.id = el.id
        if (el.class) el.node.classList.add(el.class)
    })


    // Create sprite
    const sprite = Spritz({
        picture: [
            {
                srcset: 'spritesheets/multisrc/graham-hd.png',
                media: '(min-width: 1200px)',
                width: 7800, height: 2829
            },
            {
                srcset: 'spritesheets/multisrc/graham.webp',
                width: 3900, height: 1415
            },
            {
                srcset: 'spritesheets/multisrc/graham.png',
                width: 3900, height: 1415
            }
        ],
        steps: 18,
        rows: 3
    })


    // Core
    describe('Core', () => {

        /* it('listen("click", "#id", () => {}) should add 1 eventListener', () => {
            eventt.listen('click', '#id', () => t++)
            strictEqual(eventt.events.length, 1)
        }) */

    })


    // PictureElem
    describe('PictureElem', () => {

        /* it('listen("click", "#id", () => {}) should add 1 eventListener', () => {
            eventt.listen('click', '#id', () => t++)
            strictEqual(eventt.events.length, 1)
        }) */

    })


    // API
    describe('API', () => {

        /* it('listen("click", "#id", () => {}) should add 1 eventListener', () => {
            eventt.listen('click', '#id', () => t++)
            strictEqual(eventt.events.length, 1)
        }) */

    })


    // Events
    describe('Events', () => {

        /* it('listen("click", "#id", () => {}) should add 1 eventListener', () => {
            eventt.listen('click', '#id', () => t++)
            strictEqual(eventt.events.length, 1)
        }) */

    })


})


rollup.js
const rollup = require('rollup')
const babel = require('rollup-plugin-babel')
const resolve = require('rollup-plugin-node-resolve')
const minify = require('uglify-js-harmony')
const terser = require('rollup-plugin-terser').terser

let pkg = require('./package.json')

const attribution =
`/*!
* ${ pkg.name } ${ pkg.version } - ${ pkg.description }
*
* @author       ${ pkg.author }
* @homepage     ${ pkg.homepage }
* @copyright    Copyright (c) ${ new Date().getFullYear() } ${ pkg.author }
* @license      ${ pkg.license }
* @version      ${ pkg.version }
*/
`

rollup.rollup({
    input: 'src/spritz.js',
    plugins: [
        babel(),
        resolve(),
        terser({
            output: {
                comments: (node, comment) => {
                    return (comment.type === 'comment2' && /@license/i.test(comment.value))
                }
            }
        }, minify)
    ]
}).then((bundle) => {
    bundle.write({
        file: pkg.main,
        format: 'umd',
        name: 'Spritz',
        banner: attribution,
        sourcemap: true
    })
    bundle.write({
        file: pkg.module,
        format: 'esm',
        banner: attribution,
        sourcemap: true
    })
})

src/wait.js
export default class Wait {

    constructor () {
    // initiate wait vars
        this.waitQueue = []
        this.waitTimer = false
        this.waitExecution = false
    }

    handle (func, milliseconds = false) {
    // handle wait
        this.waitQueue.push({
            'func': func,
            'timeout': milliseconds
        })

        if (!this.waitExecution) {
            this.next()
        }
    }

    next () {
    // execute next
        if (this.waitQueue.length > 0) {
            let c = this.waitQueue.shift()
            let f = c['func']
            let t = c['timeout']

            if (t !== false) {
                f()
                this.waitExecution = true
                this.waitTimer = setTimeout(() => {
                    this.next()
                }, t)
            } else {
                f()
                this.waitExecution = false
                this.next()
            }
        }
    }

}


rollup.config.js
import babel from 'rollup-plugin-babel'
import babelrc from 'babelrc-rollup'
import uglify from 'rollup-plugin-uglify'
import resolve from 'rollup-plugin-node-resolve'
import { minify } from 'uglify-js-harmony'

let pkg = require('./package.json')

const attribution =
`/*!
* ${ pkg.name } ${ pkg.version } - ${ pkg.description }
*
* @author       ${ pkg.author }
* @homepage     ${ pkg.homepage }
* @copyright    Copyright (c) ${ new Date().getFullYear() } ${ pkg.author }
* @license      ${ pkg.license }
* @version      ${ pkg.version }
*/
`

export default {
    entry: './src/spritz.js',
    plugins: [
        babel(babelrc()),
        uglify({
            output: {
                comments: (node, comment) => {
                    return (comment.type === 'comment2' && /@license/i.test(comment.value))
                }
            }
        }, minify),
        resolve({ jsnext: true, main: true })
    ],
    targets: [
        {
            dest: pkg.main,
            format: 'umd',
            exports: 'default',
            moduleName: 'Spritz',
            banner: attribution,
            sourceMap: true
        },
        {
            dest: pkg.module,
            format: 'es',
            banner: attribution,
            sourceMap: true
        }
    ]
}


gulpfile.babel.js
// imports

const pkg      = require('./package.json')
const sync     = require('browser-sync')
const del      = require('del')
const fs       = require('fs')
const gulp     = require('gulp')
const notifier = require('node-notifier')
const rollup   = require('rollup')
const babel    = require('rollup-plugin-babel')
const commonjs = require('rollup-plugin-commonjs')
const resolve  = require('rollup-plugin-node-resolve')
const eslint   = require('rollup-plugin-eslint')

// error handler

const onError = function(error) {
    notifier.notify({
        'title': 'Error',
        'message': 'Compilation failure.'
    })

    console.log(error)
}

// clean

gulp.task('clean', () => {
    return del(
        'dist/**.js',
        '!dist'
    )
})

// attribution

const attribution =
`/*!
* spritz.js ${ pkg.version } - ${ pkg.description }
* Copyright (c) ${ new Date().getFullYear() } ${ pkg.author } - ${ pkg.homepage }
* License: ${ pkg.license }
*/
`

// js

const read = {
    entry: 'src/spritz.js',
    plugins: [
        resolve({
            jsnext: true,
            main: true,
            browser: true
        }),
        eslint(),
        commonjs(),
        babel()
    ]
}

const write = {
    umd: {
        format: 'umd',
        exports: 'default',
        moduleName: 'Spritz',
        banner: attribution,
        sourceMap: true
    },
    module: {
        format: 'es',
        banner: attribution
    }
}

gulp.task('js', () => {
    return rollup
    .rollup(read)
    .then(bundle => {
        // generate UMD and ES module from bundle
        const umd = bundle.generate(write.umd)
        const module = bundle.generate(write.module)

        // write JS files
        fs.writeFileSync(pkg.main, umd.code)
        fs.writeFileSync(pkg.module, module.code)

        // write sourcemap
        fs.writeFileSync('dist/maps/spritz.js.map', umd.map.toString())
    })
    .catch(onError)
})

// server

const server = sync.create()
const reload = sync.reload

const sendMaps = (req, res, next) => {
    const filename = req.url.split('/').pop()
    const extension = filename.split('.').pop()

    if(extension === 'css' || extension === 'js') {
        res.setHeader('X-SourceMap', '/maps/' + filename + '.map')
    }

    return next()
}

const options = {
    notify: false,
    startPath: '/sandbox/',
    server: {
        baseDir: '.',
        middleware: [
            sendMaps
        ]
    },
    watchOptions: {
        ignored: '*.map'
    }
}

gulp.task('server', () => sync(options))

// watch

gulp.task('watch', () => {
    gulp.watch('src/**/*.js', ['js', reload])
})

// build and default tasks

const exists = path => {
    try {
        return fs.statSync(path).isDirectory()
    } catch(error) {}

    return false
}

gulp.task('build', ['clean'], () => {
    // create dist directories
    if(!exists('dist')) fs.mkdirSync('dist')
    if(!exists('dist/maps')) fs.mkdirSync('dist/maps')

    // run the tasks
    gulp.start('js')
})

gulp.task('default', ['build', 'server', 'watch'])


src/spritz.js
import knot from 'knot.js'
import Wait from 'wait.js'

class Spritz {

    /**
        --- CORE ---
    **/

    constructor (selector, options = {}) {
    // instance constructor
        this.options = {
            picture: options.picture || [],
            steps: options.steps || 1,
            rows: options.rows || 1,
            init: (typeof options.init !== 'undefined') ? options.init : 1,
            testunit: options.testunit || false
        }

        if (! Array.isArray(this.options.picture)) this.options.picture = [this.options.picture]

        this.selector = typeof selector === 'string'
            ? document.querySelector(selector)
            : selector

        this.emitter = knot()
        this.waitter = Wait()
        this.supportsWebP = this._supportsWebP()

        this.initiated = false

        return (this.options.init > 0)
            ? this.init(this.options.init)
            : this
    }

    _globalVars () {
    // global vars
        this.canvas = false
        this.ctx = false
        this.loaded = false
        this._resetUntil()
        this.anim = false
        this.columns = Math.ceil(this.options.steps / this.options.rows)
        this.currentFps = 15
        this.flipped = false
    }

    _throttle (callback, delay) {
    // throttle function
        let last
        let timer

        return () => {
            let context = this
            let now = +new Date()
            let args = arguments

            if (last && now < last + delay) {
                clearTimeout(timer)
                timer = setTimeout(() => {
                    last = now
                    callback.apply(context, args)
                }, delay)
            } else {
                last = now
                callback.apply(context, args)
            }
        }
    }

    _bindEvents () {
    // create events listeners
        this.resize = this._throttle(() => {
            this._resize()
        }, 250)

        window.addEventListener('resize', this.resize, false)
    }

    _unbindEvents () {
    // remove events listeners
        window.removeEventListener('resize', this.resize, false)
    }

    _resize () {
    // viewport resize triggered
        this._loadPicture()
        this.emitter.emit('resize', this.pic)
    }

    /**
        --- API ---
    **/

    init (step = 1) {
    // init the sprite
        return this.waitter.handle(this, () => {
            if (!this.initiated) {
                this.initialStep = step
                this.currentStep = step

                this._globalVars()
                this._bindEvents()
                this._createCanvas()
                this._loadPicture()

                this.initiated = true
                this.emitter.emit('ready')
            }
        })
    }

    destroy () {
    // destroy sprite instance
        return this.waitter.handle(this, () => {
            if (this.initiated) {
                // stop stuff
                this.stop()
                this._unbindEvents()

                // reset & remove canvas
                this.canvas.parentNode.removeChild(this.canvas)
                this.container.parentNode.removeChild(this.container)
                this.canvas = false
                this.ctx = false

                // turn initiated to false
                this.initiated = false

                // emitt destroy
                this.emitter.emit('destroy')

                // turn off emitters
                this.emitter.off('ready')
                this.emitter.off('destroy')
                this.emitter.off('resize')
                this.emitter.off('play')
                this.emitter.off('load')
                this.emitter.off('change')
                this.emitter.off('wait')
                this.emitter.off('flip')
                this.emitter.off('pause')
                this.emitter.off('stop')
            }
        })
    }

    play (direction = false) {
    // play animation forward
        return this.waitter.handle(this, () => {
            this.animDirection = (direction === 'backward') ? 'backward' : 'forward'
            this._startAnimation()

            this.emitter.emit('play', this.animDirection)
        })
    }

    playback () {
    // play animation backward
        return this.waitter.handle(this, () => {
            this.animDirection = 'backward'
            this._startAnimation()

            this.emitter.emit('play', this.animDirection)
        })
    }

    pause (silent = false) {
    // pause animation
        return this.waitter.handle(this, () => {
            this._pauseAnimation()

            if (!silent) {
                this.emitter.emit('pause')
            }
        })
    }

    stop () {
    // stop animation (= pause + reset)
        return this.waitter.handle(this, () => {
            this.pause(true)
            this.step(this.initialStep)

            this.emitter.emit('stop')
        })
    }

    wait (milliseconds = 0) {
    // chainable timeout
        return this.waitter.handle(this, () => {
            this.emitter.emit('wait', milliseconds)
        }, milliseconds)
    }

    step (step = 1) {
    // change the current frame/step
        return this.waitter.handle(this, () => {
            let fromStep = this.currentStep
            this.currentStep = step
            this._draw()

            this.emitter.emit('change', fromStep, this.currentStep)
        })
    }

    fps (speed) {
    // change animation speed
        return this.waitter.handle(this, () => {
            this.currentFps = speed
        })
    }

    until (step, loop = 1) {
    // next animation will stop at this
        return this.waitter.handle(this, () => {
            this.stopAtStep = step
            this.stopAtLoop = loop
        })
    }

    next () {
    // go to the next frame
        return this.waitter.handle(this, () => {
            this.animDirection = 'forward'
            let fromStep = this.currentStep
            this.currentStep = this._targetStep()
            this._draw()

            this.emitter.emit('change', fromStep, this.currentStep)
        })
    }

    prev () {
    // go to the previous frame
        return this.waitter.handle(this, () => {
            this.animDirection = 'backward'
            let fromStep = this.currentStep
            this.currentStep = this._targetStep()
            this._draw()

            this.emitter.emit('change', fromStep, this.currentStep)
        })
    }

    get (data, callback = false) {
    // return data, then call the callback function with result
        return this.waitter.handle(this, () => {
            switch (data) {
                case 'step':
                    return callback !== false ? callback.call(this, this.currentStep) : this.currentStep
                case 'picture':
                    return callback !== false ? callback.call(this, this.pic) : this.pic
                default:
                    return false
            }
        })
    }

    flip () {
    // flip the sprite horizontally
        return this.waitter.handle(this, () => {
            let css = this.flipped
                ? 'width:100%;height:100%;'
                : 'width:100%;height:100%;-webkit-transform:scale(-1, 1);-ms-transform:scale(-1, 1);transform:scale(-1, 1);-webkit-filter:FlipH;filter:FlipH;'

            this.container.setAttribute('style', css)
            this.flipped = !this.flipped
            this.emitter.emit('flip')
        })
    }

    on (...args) { return this.emitter.on(...args) }
    off (...args) { return this.emitter.off(...args) }

    /**
        --- ANIMATE ---
    **/

    _resetUntil () {
    // reset the "until()" api command
        this.stopAtLoop = false
        this.stopAtStep = false
    }

    _targetStep () {
    // return the following step to be displayed
        if (this.animDirection === 'forward') {
            return this.currentStep < this.options.steps
                ? this.currentStep + 1
                : 1
        } else {
            return this.currentStep > 1
                ? this.currentStep - 1
                : this.options.steps
        }
    }

    _animate (timestamp) {
    // frame animation
        if (this.initiated) {
            if (this.animTime === undefined) {
                this.animTime = timestamp
            }

            let seg = Math.floor((timestamp - this.animTime) / (1000 / this.currentFps))
            let pauseAnim = false

            if (seg > this.animFrame) {
                this.animFrame = seg
                let fromStep = this.currentStep
                this.currentStep = this._targetStep()

                let continueAnimate = true
                if (this.currentStep === this.stopAtStep) {
                    this.currentLoop ++
                    if (this.currentLoop === this.stopAtLoop) {
                        continueAnimate = false
                    }
                }

                if (continueAnimate) {
                    this._draw()
                    this.emitter.emit('change', fromStep, this.currentStep)
                } else {
                    this._draw()
                    this.pause()
                    pauseAnim = true
                }
            }

            if (!pauseAnim && !this.requireStop) {
                this.anim = window.requestAnimationFrame((timestamp) => this._animate(timestamp))
            }
        }
    }

    _startAnimation () {
    // start animation
        if (!this.anim) {
            this.requireStop = false
            this.animTime = undefined
            this.animFrame = 0
            this.currentLoop = 0
            this.anim = window.requestAnimationFrame((timestamp) => this._animate(timestamp))
        } else {
            this.pause()
            this.play(this.animDirection)
        }
    }

    _pauseAnimation () {
    // pause animation
        if (this.anim) {
            this._resetUntil()
            this.requireStop = true
            window.cancelAnimationFrame(this.anim)
            this.anim = false
        }
    }

    /**
        --- DETECT & CALCULATE ---
    **/

    _selectPicture () {
    // select picture src from list
        for (let i = 0; i < this.options.picture.length; i++) {
            let pic = this.options.picture[i]
            if (this._supportsFormat(pic.srcset) && this._matchesMedia(pic.media)) {
                this.pic = pic
                return pic.srcset
            }
        }
        return false
    }

    _supportsFormat (filename) {
    // return true if filename is a supported format
        let ext = this._getExtension(filename)
        return (ext === 'webp' && this.supportsWebP) || ext !== 'webp'
    }

    _supportsWebP () {
    // return true if webP is supported
        let canvas = document.createElement('canvas')
        canvas.width = canvas.height = 1
        return canvas.toDataURL && canvas.toDataURL('image/webp') && canvas.toDataURL('image/webp').indexOf('data:image/webp') === 0
    }

    _getExtension (filename) {
    // return the filename extension
        return (/[.]/.exec(filename)) ? /[^.]+$/.exec(filename)[0] : undefined
    }

    _matchesMedia (query = undefined) {
    // return true if matches the media query
        let mq = window.matchMedia(query)
        return query !== undefined ? mq.matches : true
    }

    _setDimensions () {
    // calculate sprite dimensions
        this.stepWidth = this.pic.width / this.columns
        this.stepHeight = this.pic.height / this.options.rows
        this.stepRatio = this.stepWidth / this.stepHeight

        this.parentWidth = this.selector.clientWidth
        this.parentHeight = this.selector.clientHeight
        this.parentRatio = this.parentWidth / this.parentHeight

        if (this.pic.objectFit === 'cover') {
            if (this.stepRatio >= this.parentRatio) {
                this.canvasHeight = this.parentHeight
                this.canvasWidth = (this.stepWidth * this.canvasHeight) / this.stepHeight
            } else {
                this.canvasWidth = this.parentWidth
                this.canvasHeight = (this.stepHeight * this.canvasWidth) / this.stepWidth
            }
        } else {
            if (this.stepRatio >= this.parentRatio) {
                this.canvasWidth = this.parentWidth
                this.canvasHeight = (this.stepHeight * this.canvasWidth) / this.stepWidth
            } else {
                this.canvasHeight = this.parentHeight
                this.canvasWidth = (this.stepWidth * this.canvasHeight) / this.stepHeight
            }
        }

        this.canvas.width = this.canvasWidth
        this.canvas.height = this.canvasHeight
    }

    /**
        --- CREATE & DRAW ---
    **/

    _loadPicture () {
    // load source picture
        this.picture = new Image()
        this.picture.onload = () => {
            if (!this.pic.width) this.pic.width = this.picture.naturalWidth
            if (!this.pic.height) this.pic.height = this.picture.naturalHeight

            if (!this.loaded) {
                this.emitter.emit('load', this.pic)
                this.loaded = true
            }
            this._draw()
        }
        this.picture.src = this._selectPicture()
    }

    _draw () {
    // draw sprite
        if (this.initiated) {
            this._setDimensions()
            this._drawPicture()
        }
    }

    _drawPicture () {
    // draw picture into canvas
        let targetColumn = Math.floor((this.currentStep - 1) % this.columns)
        let targetRow = Math.floor((this.currentStep - 1) / this.columns)

        let posX = targetColumn * this.stepWidth
        let posY = targetRow * this.stepHeight

        if (!this.options.testunit) {
            this.ctx.clearRect(0, 0, this.canvasWidth, this.canvasHeight)

            this.ctx.drawImage(
                this.picture,
                Math.round(posX), Math.round(posY),
                this.stepWidth,
                this.stepHeight,
                0, 0,
                this.canvasWidth,
                this.canvasHeight
            )
        }
    }

    _createCanvas () {
    // create html5 canvas
        this.canvas = document.createElement('canvas')
        this.canvas.setAttribute('style', 'position:absolute;left:50%;top:50%;-webkit-transform:translateY(-50%) translateY(1px) translateX(-50%) translateX(1px);-ms-transform:translateY(-50%) translateY(1px) translateX(-50%) translateX(1px);transform:translateY(-50%) translateY(1px) translateX(-50%) translateX(1px);')

        this.container = document.createElement('div')
        this.container.setAttribute('style', 'width:100%;height:100%;position:relative;')
        this.container.appendChild(this.canvas)

        this.selector.appendChild(this.container)
        this.ctx = this.canvas.getContext('2d')
    }

}

export default (...args) => { return new Spritz(...args) }
